bats/tests/k8s/helm-install-rancher.bats

```diff
@@ -98,22 +163,65 @@ deploy_rancher() {
+    try assert_not_empty_list helm list --namespace cert-manager --deployed --output json --selector name=cert-manager
+    wait_for_kube_deployment_available --namespace cert-manager cert-manager
 
     local host
     host=$(traefik_hostname) || return
 
     comment "Installing rancher $rancher_chart_version"
+    # The helm install can take a long time, especially on CI
     helm upgrade \
         --install rancher rancher-latest/rancher \
         --version "$rancher_chart_version" \
         --namespace cattle-system \
         --set hostname="$host" \
-        --wait \
-        --timeout=10m \
+        --set replicas=1 \
         --create-namespace
+    try assert_not_empty_list helm list --all --output json --namespace cattle-system --selector name=rancher
+    try assert_not_empty_list helm list --deployed --output json --namespace cattle-system --selector name=rancher
+    try kubectl get ingress --namespace cattle-system rancher
+    try assert_not_empty_list kubectl get ingress --namespace cattle-system rancher --output jsonpath='{.status.loadBalancer.ingress}'
+
+    try --max 60 --delay 10 kubectl get namespace fleet-local
+    try --max 120 --delay 10 kubectl get namespace local
+    try --max 120 --delay 10 kubectl get namespace cattle-global-data
+    try --max 60 --delay 10 kubectl get namespace fleet-default
+
+    try assert_not_empty_list kubectl get pods --namespace cattle-system --selector app=rancher --output jsonpath='{.items}'
+
+    # Unfortunately, the Rancher pod could get restarted, so we need to put this in a loop :(
+    try wait_for_rancher_pod
+
+    try --max 60 --delay 10 assert_true kubectl get APIServices v3.project.cattle.io --output=jsonpath='{.status.conditions[?(@.type=="Available")].status}'
+
+    try --max 60 --delay 10 kubectl get namespace cattle-fleet-system
+    try --max 60 --delay 10 kubectl get namespace cattle-system
+
+    try --max 60 --delay 10 kubectl get deployment --namespace cattle-fleet-system fleet-controller
+    try assert_kube_deployment_available --namespace cattle-fleet-system gitjob
+    try assert_kube_deployment_available --namespace cattle-fleet-system fleet-controller
+
+    try --max 120 --delay 10 assert_not_empty_list kubectl get pods --namespace cattle-system --selector app=rancher-webhook --output jsonpath='{.items}'
+
+    # Unfortunately, the webhook pod might restart too :(
+    try wait_for_webhook_pod
+
+    try --max 60 --delay 10 assert_kube_deployment_available --namespace cattle-system rancher
+    try --max 60 --delay 10 assert_kube_deployment_available --namespace cattle-fleet-local-system fleet-agent
+    try --max 120 --delay 10 assert_kube_deployment_available --namespace cattle-system rancher-webhook
+
+    # The rancher pod sometimes falls over on its own; retry in a loop
+    local i
+    for i in {1..10}; do
+        sleep 1
+        try --max 60 --delay 10 assert_kube_deployment_available --namespace cattle-system rancher
+    done
```

Comment:
I don't get the purpose of the loop. It checks that the deployment is available at least 10 times, but it can fall over up to 59 times between each check. So what does this actually prove?

Reply:
That was my point: how does it mean it has stopped flapping. It just means you have observed it 10 times when it was running, but that doesn't mean it has stopped flapping. Because you continue to retry until it is up again.

The app can be down for almost 100 minutes. As long as the app is temporarily up for at 10 seconds within every 10 minutes interval, the test will pass even though the app can be down 98 out of those 100 minutes.